Sources/LeafKit/LeafRenderer.swift

```diff
@@ -1,59 +1,145 @@
     func insert(
         _ document: ResolvedDocument,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument>
+    
+    func insert(
+        _ document: ResolvedDocument,
+        on loop: EventLoop,
+        replace: Bool
+    ) -> EventLoopFuture<ResolvedDocument>
+    
     func load(
         documentName: String,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument?>
+    
+    /// - return nil if cache entry didn't exist in the first place, true if purged
+    /// - will never return false in this design but should be capable of it
+    ///   in the event a cache implements dependency tracking between templates
+    func remove(
+        _ documentName: String,
+        on loop: EventLoop
+    ) -> EventLoopFuture<Bool?>
+    
+    func entryCount() -> Int
+    
     var isEnabled : Bool { get set }
 }
 
 public final class DefaultLeafCache: LeafCache {
     let lock: Lock
     var cache: [String: ResolvedDocument]
     public var isEnabled: Bool = true
-    
+
     public init() {
         self.lock = .init()
         self.cache = [:]
     }
     
+    // Superseded by insert with remove: parameter - Remove in Leaf-Kit 2?
     public func insert(
         _ document: ResolvedDocument,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument> {
+        self.insert(document, on: loop, replace: false)
+    }
+    
+    public func insert(
+        _ document: ResolvedDocument,
+        on loop: EventLoop,
+        replace: Bool = false
+    ) -> EventLoopFuture<ResolvedDocument> {
+        // future fails if caching is enabled
+        guard isEnabled else { return loop.makeSucceededFuture(document) }
+        
         self.lock.lock()
         defer { self.lock.unlock() }
-       if isEnabled {
-            self.cache[document.name] = document
+        // return an error if replace is false and the document name is already in cache
+        switch (self.cache.keys.contains(document.name),replace) {
+            case (true, false): return loop.makeFailedFuture(LeafError.keyExists(document.name))
+            default: self.cache[document.name] = document
         }
         return loop.makeSucceededFuture(document)
     }
-    
+
     public func load(
         documentName: String,
         on loop: EventLoop
     ) -> EventLoopFuture<ResolvedDocument?> {
+        guard isEnabled == true else { return loop.makeSucceededFuture(nil) }
         self.lock.lock()
         defer { self.lock.unlock() }
         return loop.makeSucceededFuture(self.cache[documentName])
     }
+    
+    public func remove(
+        _ documentName: String,
+        on loop: EventLoop
+    ) -> EventLoopFuture<Bool?> {
+        guard isEnabled == true else { return loop.makeFailedFuture(LeafError.cachingDisabled) }
+        
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        guard self.cache[documentName] != nil else { return loop.makeSucceededFuture(nil) }
```

Comment:
Yeah - that's by design, as in:
True: The key was removed
False: The key could not be removed (not yet returned)
Nil: Indicates nothing was removed because nothing was there

Meaning that, once I hopefully add graph dependency for cascading re-resolutions of non-flat ASTs, removing may be BLOCKED because others won't allow it (eg, a false result).

Reply:
Sure, I'm just saying using `removeValue(forKey:)` is a neater solution 🙂

Reply:
Agreed FOR NOW that would work cleanly but I'd have to change it back to testing for existence first before deciding WHETHER to allow removal, so it more neatly fits this function's definition.